File: combinatorics/resource/ST/project_5/bai_toan_4/trees/graph/graph_to_ap.cpp
#include <iostream>
#include <vector>
#include <sstream>
using namespace std;

void dfs(int u, int p, const vector<vector<int>>& adj, vector<int>& parent) {
    parent[u] = p;
    for (int v : adj[u]) {
        if (v != p) {
            dfs(v, u, adj, parent);
        }
    }
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;
    cin.ignore();

    vector<vector<int>> adj(n);
    cout << "Nhap danh sach ke:\n";
    for (int i = 0; i < n; ++i) {
        cout << i << ": ";
        string line;
        getline(cin, line);
        stringstream ss(line);
        int v;
        while (ss >> v)
            adj[i].push_back(v);
    }

    vector<int> parent(n, -1);
    dfs(0, -1, adj, parent);  // Goc la 0

    cout << "Mang parent:\n";
    for (int p : parent) cout << p << " ";
    cout << "\n";

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/trees/graph/graph_to_ap.py
def dfs(u, p, adj, parent):
    parent[u] = p  # Gan cha cua u la p
    for v in adj[u]:
        if v != p:
            dfs(v, u, adj, parent)

n = int(input("Nhap so dinh: "))
adj = [[] for _ in range(n)]

print("Nhap danh sach ke:")
for i in range(n):
    line = list(map(int, input(f"{i}: ").split()))
    adj[i].extend(line)

parent = [-1] * n
dfs(0, -1, adj, parent)  # Goc la 0

print("Mang parent:")
print(parent)